backends/xsmm/ceed-xsmm-tensor-f64.c

```diff
@@ -7,148 +7,65 @@
 
 #include <ceed.h>
 #include <ceed/backend.h>
-#include <ceed/hash.h>
-#include <ceed/khash.h>
 #include <libxsmm.h>
-#include <stddef.h>
 
 #include "ceed-xsmm.h"
 
-//------------------------------------------------------------------------------
-// Tensor Contract C=1
-//------------------------------------------------------------------------------
-static int CeedTensorContract_Xsmm_C1(CeedTensorContract contract, CeedInt A, CeedInt B, CeedInt C, CeedInt J, const double *restrict t,
-                                      CeedTransposeMode t_mode, const CeedInt add, const double *restrict u, double *restrict v) {
-  double alpha = 1.0, beta = 1.0;
-  char   trans_u = 'N', trans_t = 'N';
-  if ((t_mode == CEED_TRANSPOSE && C != 1) || (t_mode == CEED_NOTRANSPOSE && C == 1)) trans_t = 'T';
-
-  if (!add) beta = 0.0;
-
-  // libXSMM GEMM
-  libxsmm_dgemm(&trans_t, &trans_u, &J, &A, &B, &alpha, &t[0], NULL, &u[0], NULL, &beta, &v[0], NULL);
-
-  return CEED_ERROR_SUCCESS;
-}
-
 //------------------------------------------------------------------------------
 // Tensor Contract Apply
 //------------------------------------------------------------------------------
 static int CeedTensorContractApply_Xsmm(CeedTensorContract contract, CeedInt A, CeedInt B, CeedInt C, CeedInt J, const double *restrict t,
                                         CeedTransposeMode t_mode, const CeedInt add, const double *restrict u, double *restrict v) {
-  CeedTensorContract_Xsmm *impl;
-  CeedCallBackend(CeedTensorContractGetData(contract, &impl));
-
-  // Get kernel
-  libxsmm_dmmfunction kernel;
-  CeedHashIJKLMKey    key = {B, C, J, t_mode, add};
-  khint_t             k   = kh_get(f64, impl->lookup_f64, key);
-  CeedHashGetValue(impl->lookup_f64, k, kernel);
+  Ceed ceed;
+  CeedCallBackend(CeedTensorContractGetCeed(contract, &ceed));
 
-  // Run kernel or fallback to default implementation
-  if (C != 1) {
-    for (CeedInt a = 0; a < A; a++) LIBXSMM_MMFUNCTION_KERNEL(&u[a * B * C], &t[0], &v[a * J * C]);
+  if (C == 1) {
+    // Build or query the required kernel
+    const int                  flags_t    = LIBXSMM_GEMM_FLAGS(!t_mode ? 'T' : 'N', 'N');
+    const int                  flags_ab   = (!add) ? LIBXSMM_GEMM_FLAG_BETA_0 : 0;
+    const int                  flags      = (flags_t | flags_ab);
+    const libxsmm_gemm_shape   gemm_shape = libxsmm_create_gemm_shape(J, A, B, !t_mode ? B : J, B, J, LIBXSMM_DATATYPE_F64, LIBXSMM_DATATYPE_F64,
```

Comment:
Can we combine the two `ceed-xsmm-tensor` files now since the only difference is the `LIBXSMM_DATATYPE_*`, which we can select based upon the value of `CEED_SCALAR_TYPE`?

Reply:
Nice, see [fe99f24](https://github.com/CEED/libCEED/pull/1248/commits/fe99f2494c1cc00aebca1d5c9b9672a3607e535b).

Reply:
I have a similar update for the AVX backend. Can I add this here or should I make a new PR?